platform/linux-generic/odp_packet_io.c

```diff
@@ -1577,10 +1577,29 @@ int odp_pktin_recv_tmo(odp_pktin_queue_t queue, odp_packet_t packets[], int num,
 	odp_time_t t1, t2;
 	struct timespec ts;
 	int started = 0;
+	pktio_entry_t *entry;
 
 	ts.tv_sec  = 0;
 	ts.tv_nsec = SLEEP_NSEC;
 
+	entry = get_pktio_entry(queue.pktio);
+	if (entry == NULL) {
+		ODP_DBG("pktio entry %d does not exist\n", queue.pktio);
+		return -1;
+	}
+
+	if (entry->s.ops->recv_tmo) {
+		uint64_t usecs = (wait * SLEEP_NSEC + 999) / 1000;
```

Comment:
The motivation for a separate `odp_pktin_wait_time()` function was to avoid having to do time conversions in the fast path functions. As `odp_pktin_wait_time()` already returns the time in microseconds no additional conversions are needed here. Same thing for `odp_pktin_recv_mq_tmo()`.

Currently, the value returned by `odp_pktin_recv_mq_tmo()` depends on `SLEEP_NSEC` definition. This should be modified so that the returned value is always in microseconds.

Reply:
I'm somewhat worried that somebody could adjust SLEEP_NSEC and then the entire thing fails. What about this:

<pre>
#if SLEEP_NSEC == 1000
  uint64_t usecs = wait;
#else
  uint64_t usecs = (wait * SLEEP_NSEC + 999) / 1000;
#endif
</pre>

For the return value, you mean <code>odp_pktin_wait_time()</code> and not <code>odp_pktin_recv_mq_tmo()</code>, right? Because I don't see how the return value of the latter could depend on <code>SLEEP_NSEC</code> definition.

Also, if we modify <code>odp_pktin_wait_time()</code> to always return microseconds, then <code>odp_pktin_recv_mq_tmo()</code> suddenly starts to work incorrectly with <code>SLEEP_NSEC != 1000</code> as there is a <code>wait--</code> line in there that does this just before sleeping one microsecond, exactly.

Reply:
Ifdefs are not allowed.  IMO the cleanest fix would be to add a separate commit (before this one) to this patch set, which modifies odp_pktin_wait_time() to always return microseconds and converts SLEEP_NSEC to SLEEP_USEC (and does necessary changes to make everything work).

Or does someone oppose having a minimum sleep time of 1us between pktin receive calls in the old code path?

Reply:
Now that's a great idea! If somebody sees one microsecond as too big sleep time, it is always possible to implement a "divisor": multiply the sleep time by e.g. 4, and sleep e.g. 250 nanoseconds. No slow division operation is needed for this "divisor", only multiplication. But I won't implement this divisor now, as SLEEP_NSEC anyway always used to be 1000.

Reply:
Agreed, the "divisor" can be added later if needed.